Back-off logic for indexers

Taloth opened the pull request:
<s>Work in progress, but a request for comments.
Won't actually run, didn't actually add a db migration yet.

I'm not happy with how this is done architecturally.</s>

The IndexerStatusService seems okay. We could also do this in IndexerBase directly but the service seems a better option.
The IndexerStatus contains both back-off logic and last recent-feed tracking. Two separate features, but wanted to look at both coz it affects the architecture.

The back-off logic should be self-explanatory. Upon failure the indexer will not be used for a while, the duration will exponentially increase after every failure up to 24h. Every success will reduce the duration again.
So if you get Failure, Failure, Failure, Success, Failure. Then the Success won't actually completely reset the escalation.

The recent-feed tracking is intended to be able to grab multiple pages of the recent feed to catch up in case of a server or client down-time (non-24/7 operation). The status is used to determine how far back it needs to go and whether there's still a gap left.
The tracking should be okay (no tests yet), but the gaps aren't reported yet.
That should probably be done by event.
I also didn't increase the number of allowed pages for the recent feed. But once we do we might be able to run Sonarr in non-24/7 mode, should make more than a few ppl happy.

Several problems exist:
1) <s>Since the nzb/torrent is grabbed in the DownloadClient instance, it's more difficult to differentiate between exceptions thrown by the nzb/torrent download and exceptions thrown by the download client. Atm I catch the exceptions in the caller (DownloadService).</s> done
2) <s>HttpIndexerBase currently gets a reference to IIndexerStatusService, I'm thinking we might want to use events instead.</s> nope
3) <s>Recent feed gaps aren't yet reported, we could raise an event, but we first have to run through all the indexers. So not sure how to do that properly. Could also just store the gap in the IndexerStatus and reset it when the MissingEpisode search finishes.</s> done
4) <s>We need to add a HealthCheck to warn about it.</s> done

markus101 commented:
1. Exceptions in DownloadClient exception is something I wanted to handle so we could also deal with the download client being down (and storing the nzb) - perhaps we can raise different types of exceptions?

2. That would cut down on the indexer dependencies

3. I think this depends on whether we want to check via RSS or if we are okay with relying on a search. I think I'm leaning towards the former since there are non-search indexers.

4. Do we want these to be dismissable? Maybe we should track them in the DB?

Taloth commented:
1) Yes, but atm we differentiate between DownloadClientException (something the client reports to us) and other exceptions (for example HttpException/WebException when the connection fails). Atm there's no way of differentiating between WebException by indexer or WebException by downloadclient, that's fixable of course, with a well placed try catch. We should probably throw any nzb/torrent download failure as IndexerException or something with the actual exception as InnerException.

2) Not really, we would be replacing IIndexerStatusService with IEventAggregator. It will be more loosely coupled with events.

3) I'm not okay with relying on a search. That's why I'm doing this, to be able to fill in the blanks for a longer period and only resort to search if that fails. I'm leaning towards storing the gap in the IndexerStatus as well.

4) No, you can dismiss it by disabling the indexer. We already track the back-off in the database, so it will be easy possible to reraise the healthcheck on restart.

Taloth commented:
@markus101 
Okay, rewritten most of this. I didn't use eventAggregrator coz in most cases the caller service had to retrieve the IndexerStatus as well.

Also added HealthCheck to get a more prominent warning out there in case of failures.

Once concern is the current escalation speed.
5 sec, 10 sec, 20 sec, 40 sec, 80 sec, 160 sec, 320 sec, 640 sec etc. This means it takes 9 failures in a row to reach a 30min back-off.
I could start at 30 sec, in which case it will reach it in 6.
or start at 5 minutes, in which case it will reach it in 4.
The starting point isn't that important, coz it will generally take 15 min anyway to get to the next rss sync, but on one end you want it to get disabled fast enough, but not escalate to 24h delay too fast.

If I start at 5 minutes, it will start skipping rss syncs after 3 failures, and reach the 24h back-off after 9 failures (roughly 40h after the first failure).

So feel free to shoot at it now. Atm I'm running it live to see if the escalation works as intended.

kayone commented on src/NzbDrone.Core/Datastore/Migration/091_added_indexerstatus.cs:
```diff
@@ -0,0 +1,26 @@
+                  .WithColumn("LastFailure").AsDateTime().Nullable()
+                  .WithColumn("FailureEscalation").AsInt32().NotNullable()
+                  .WithColumn("BackOffDate").AsDateTime().Nullable()
+                  .WithColumn("LastRecentSearch").AsDateTime().Nullable()
```
`MostRecentSearch`
`MostRecentReleaseInfo`

Taloth replied:
It's actually MostRecent-RecentSearch. Recent = RssSync. It's not actually a Search.

kayone commented on src/NzbDrone.Core/Indexers/IndexerStatus.cs:
```diff
@@ -0,0 +1,21 @@
+        public int IndexerId { get; set; }
+
+        public DateTime? FirstFailure { get; set; }
+        public DateTime? LastFailure { get; set; }
```
`MostRecentFailure`

Taloth replied:
FirstFailure = first failure in the current escalation cycle..
LastFailure = last failure in the current escalation cycle.

I can change LastFailure to MostRecentFailure, but then what should FirstFailure be renamed to?

kayone commented on src/NzbDrone.Core/Indexers/IndexerStatus.cs:
```diff
@@ -0,0 +1,21 @@
+
+        public DateTime? FirstFailure { get; set; }
+        public DateTime? LastFailure { get; set; }
+        public int FailureEscalation { get; set; }
```
`EscalationLevel` or `EscalationCount`

kayone commented on src/NzbDrone.Core/Indexers/IndexerStatusService.cs:
```diff
@@ -0,0 +1,117 @@
+            }
+        }
+
+        public void UpdateRecentSearchStatus(int indexerId, ReleaseInfo releaseInfo, bool fullyUpdated)
```
shouldn't this be `UpdateRssStatus`, since you are updating the rss sync flag.

Taloth replied:
Yes, renamed the property earlier, forgot the method.

kayone commented on src/NzbDrone.Core/Indexers/RssSyncService.cs:
```diff
@@ -44,7 +51,7 @@ private ProcessedDecisions Sync()
             _logger.ProgressInfo("Starting RSS Sync");
 
-            var reports = _rssFetcherAndParser.Fetch().Concat(_pendingReleaseService.GetPending()).ToList();
+            var reports = _rssFetcherAndParser.Fetch().Concat(FilterByIndexer(_pendingReleaseService.GetPending())).ToList();
```
there is way too many method calls in one line, I would break them into different calls with proper variable name for what they are returning.

kayone replied:
I would suggest `var pendingReleases` and `var liveReports` and concating them into `reports`.

also it's a bit strange that  `RssFetcherAndParser` is dealing with indexer backoff, but `PendingReleaseService` doesn't. it means `RssSyncService` knows a bit too much about their internals.

kayone commented on src/NzbDrone.Core/Indexers/RssSyncService.cs:
```diff
@@ -60,15 +67,70 @@ private ProcessedDecisions Sync()
+            foreach (var release in releases)
+            {
+                bool ignore;
+                if (!indexerBackOff.TryGetValue(release.IndexerId, out ignore))
```
you should use `Cached`instead of doing your own cache. that's exactly why it's there.

Taloth replied:
It's a very short-lived cache. I could instantiate Cached, but I shouldn't use CacheManager, if that's what you wanted.

kayone replied:
Yeah. Just create a cached everytime

kayone commented on src/NzbDrone.Core/Indexers/RssSyncService.cs:
```diff
@@ -60,15 +67,70 @@ private ProcessedDecisions Sync()
             return processed;
         }
 
+        private IEnumerable<ReleaseInfo> FilterByIndexer(IEnumerable<ReleaseInfo> releases)
```
`FilterBlockedIndexers`

kayone commented on src/NzbDrone.Core/Indexers/RssSyncService.cs:
```diff
@@ -60,15 +67,70 @@ private ProcessedDecisions Sync()
+                {
+                    var indexerStatus = _indexerStatusService.GetIndexerStatus(release.IndexerId);
+
+                    if (indexerStatus == null || !indexerStatus.DisabledTill.HasValue || indexerStatus.DisabledTill.Value < DateTime.UtcNow)
```
should never be null is you are defaulting in `IndexerStatusService`

kayone commented on src/NzbDrone.Core/Indexers/RssSyncService.cs:
```diff
@@ -60,15 +67,70 @@ private ProcessedDecisions Sync()
             {
-                _logger.Info("RSS Sync hasn't run since: {0}. Searching for any missing episodes since then.", message.LastExecutionTime.Value);
-                _episodeSearchService.MissingEpisodesAiredAfter(message.LastExecutionTime.Value.AddDays(-1), grabbedOrPending.SelectMany(d => d.RemoteEpisode.Episodes).Select(e => e.Id));
+                var lastGap = rssStarted;
```
to be honest, I don't think this should run as part of RSS sync. it should be it's own command.

kayone commented on src/NzbDrone.Core/Download/TorrentClientBase.cs:
```diff
@@ -126,7 +126,13 @@ private string DownloadFromWebUrl(RemoteEpisode remoteEpisode, String torrentUrl
             }
+            catch (HttpException ex)
+            {
+                _logger.ErrorException(String.Format("Downloading torrentfile for episode '{0}' failed ({1})",
```
`torrent file` or `.torrent`

kayone commented on src/NzbDrone.Core/Download/UsenetClientBase.cs:
```diff
@@ -49,6 +49,13 @@ public override String Download(RemoteEpisode remoteEpisode)
             }
+            catch (HttpException ex)
+            {
+                _logger.ErrorException(String.Format("Downloading nzb for episode '{0}' failed ({1})",
```
can't we move this to download service since both `torrentClient` and `usenet` client do it?

Taloth replied:
No, coz torrentclient will use magnet.

kayone commented on src/NzbDrone.Core/HealthCheck/Checks/IndexerStatusCheck.cs:
```diff
@@ -0,0 +1,47 @@
+        {
+            var enabledIndexers = _indexerFactory.GetAvailableProviders();
+            var backOffIndexers = enabledIndexers
+                .Select(v => Tuple.Create(v, _indexerStatusService.GetIndexerStatus(v.Definition.Id)))
```
use an anonymous type instead of a tuple. makes things easier to read. also break this in multiple stages with proper variable names.

kayone commented on src/NzbDrone.Core/HealthCheck/Checks/IndexerStatusCheck.cs:
```diff
@@ -0,0 +1,47 @@
+
+            if (backOffIndexers.Count > 1)
+            {
+                return new HealthCheck(GetType(), HealthCheckResult.Warning, "Multiple indexers unusable due to failures", "#indexer-unusable-due-to-failures");
```
use actual count instead of "multiple"

kayone commented on src/NzbDrone.Core/Housekeeping/Housekeepers/CleanupOrphanedIndexerStatus.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+            var mapper = _database.GetDataMapper();
+
+            mapper.ExecuteNonQuery(@"DELETE FROM IndexerStatus
```
can we not just delete the status when someone deletes an indexer instead of relying on a housekeeping?

Taloth replied:
That involves modifying and overriding Delete on IndexerFactory and have IndexerFactory IIndexerStatusRepository as well. Fine by me though.

kayone replied:
Event?

kayone commented on src/NzbDrone.Core/IndexerSearch/NzbSearchService.cs:
```diff
@@ -257,6 +260,13 @@ private List<DownloadDecision> Dispatch(Func<IIndexer, IEnumerable<ReleaseInfo>>
+                var indexerStatus = _indexerStatusService.GetIndexerStatus(indexer.Definition.Id);
+                if (indexerStatus != null && indexerStatus.DisabledTill.HasValue && indexerStatus.DisabledTill.Value > DateTime.UtcNow)
+                {
+                    _logger.Debug("Temporarily ignoring indexer {0} till {1} due to recent failures.", indexer.Definition.Name, indexerStatus.DisabledTill.Value);
```
maybe as info? users should be able to tell why an indexer isn't being used.

Taloth replied:
They already have warnings of the earlier errors. If you do this as Info you get alot of entries saying the same thing over and over again.

kayone commented on src/NzbDrone.Core/IndexerSearch/NzbSearchService.cs:
```diff
@@ -257,6 +260,13 @@ private List<DownloadDecision> Dispatch(Func<IIndexer, IEnumerable<ReleaseInfo>>
             foreach (var indexer in indexers)
             {
+                var indexerStatus = _indexerStatusService.GetIndexerStatus(indexer.Definition.Id);
+                if (indexerStatus != null && indexerStatus.DisabledTill.HasValue && indexerStatus.DisabledTill.Value > DateTime.UtcNow)
```
there is too much logic around whether or not an indexer is usable due to it's status, I think we should move this logic to `IndexerStatus` class as methods. `indexerStatus.IsDisabled()`

Taloth replied:
Agreed

kayone commented on src/NzbDrone.Core/Indexers/FetchAndParseRssService.cs:
```diff
@@ -42,6 +45,13 @@ public List<ReleaseInfo> Fetch()
             foreach (var indexer in indexers)
             {
+                var indexerStatus = _indexerStatusService.GetIndexerStatus(indexer.Definition.Id);
+                if (indexerStatus != null && indexerStatus.DisabledTill.HasValue && indexerStatus.DisabledTill.Value > DateTime.UtcNow)
```
maybe this should be moved to `indexerFactory.RssEnabled()` with a boolean flag `bool includedBlocked`

kayone commented on src/NzbDrone.Core/Indexers/HttpIndexerBase.cs:
```diff
@@ -132,7 +143,38 @@ protected virtual IList<ReleaseInfo> FetchReleases(IList<IEnumerable<IndexerRequ
+                        }
+                        else
+                        {
+                            if (pagedReleases.Count >= MaxNumResultsPerQuery)
```
`else if`

kayone commented on src/NzbDrone.Core/Indexers/HttpIndexerBase.cs:
```diff
@@ -132,7 +143,38 @@ protected virtual IList<ReleaseInfo> FetchReleases(IList<IEnumerable<IndexerRequ
+
+                            if (pagedReleases.Count >= MaxNumResultsPerQuery)
+                            {
+                                if (oldestDate < DateTime.UtcNow - TimeSpan.FromHours(24))
```
merge these two ifs,

kayone commented on src/NzbDrone.Core/Indexers/HttpIndexerBase.cs:
```diff
@@ -120,6 +120,17 @@ protected virtual IList<ReleaseInfo> FetchReleases(IList<IEnumerable<IndexerRequ
+                if (isRecent)
+                {
+                    var indexerStatus = _indexerStatusService.GetIndexerStatus(Definition.Id);
+                    if (indexerStatus != null)
```
don't need null gaurd.

kayone commented on src/NzbDrone.Core/Indexers/HttpIndexerBase.cs:
```diff
@@ -140,44 +182,61 @@ protected virtual IList<ReleaseInfo> FetchReleases(IList<IEnumerable<IndexerRequ
                 if (webException.Message.Contains("502") || webException.Message.Contains("503") ||
                     webException.Message.Contains("timed out"))
                 {
+                    _indexerStatusService.ReportFailure(Definition.Id);
```
why not just do it outside of the if instead of in both if and the else?

kayone commented on src/NzbDrone.Common/Http/TooManyRequestsException.cs:
```diff
@@ -0,0 +1,21 @@
+        public TooManyRequestsException(HttpRequest request, HttpResponse response)
+            : base(request, response)
+        {
+            if ((response.Headers.ContainsKey("Retry-After")))
```
extra `()`

kayone commented on src/NzbDrone.Core/Download/DownloadService.cs:
```diff
@@ -54,7 +60,26 @@ public void DownloadReport(RemoteEpisode remoteEpisode)
+            try
+            {
+                downloadClientId = downloadClient.Download(remoteEpisode);
+                _indexerStatusService.ReportSuccess(remoteEpisode.Release.IndexerId);
```
I think `RecordSuccess` might be a more accurate name, not a big deal.

kayone commented on src/NzbDrone.Core/Download/DownloadService.cs:
```diff
@@ -54,7 +60,26 @@ public void DownloadReport(RemoteEpisode remoteEpisode)
+                var http429 = ex.InnerException as TooManyRequestsException;
+                if (http429 != null)
+                {
+                    _indexerStatusService.ReportFailure(remoteEpisode.Release.IndexerId, http429.RetryAfter);
```
same here,

kayone commented on src/NzbDrone.Core/Download/DownloadService.cs:
```diff
@@ -54,7 +60,26 @@ public void DownloadReport(RemoteEpisode remoteEpisode)
+            }
+            catch (ReleaseDownloadException ex)
+            {
+                var http429 = ex.InnerException as TooManyRequestsException;
```
I think having a specific catch is cleaner than relaying on casting and `if`.

Taloth replied:
The original TooManyRequests is caught (all exceptions are caught) and thrown as ReleaseDownloadException. That's why I'm checking the InnerException.

kayone commented on src/NzbDrone.Core/IndexerSearch/NzbSearchService.cs:
```diff
@@ -25,6 +25,7 @@ public interface ISearchForNzb
     public class NzbSearchService : ISearchForNzb
     {
         private readonly IIndexerFactory _indexerFactory;
+        private readonly IIndexerStatusService _indexerStatusService;
```
I don't think it's being used anymore

kayone commented on src/NzbDrone.Core/Indexers/HttpIndexerBase.cs:
```diff
@@ -132,7 +143,33 @@ protected virtual IList<ReleaseInfo> FetchReleases(IList<IEnumerable<IndexerRequ
+                                fullyUpdated = true;
+                                break;
+                            }
+                            var oldestDate = page.Select(v => v.PublishDate).Min();
```
`oldestReleaseDate`

kayone commented on src/NzbDrone.Core/Indexers/IndexerFactory.cs:
```diff
@@ -50,12 +50,35 @@ public override IndexerDefinition GetProviderCharacteristics(IIndexer provider,
+        {
+            foreach (var indexer in indexers)
+            {
+                var indexerStatus = _indexerStatusService.GetIndexerStatus(indexer.Definition.Id);
```
use `GetBlockedIndexers()` single call to DB instead of one call per indexer.

kayone commented on src/NzbDrone.Core/Indexers/IndexerStatusService.cs:
```diff
@@ -0,0 +1,135 @@
+
+        public IndexerStatus GetIndexerStatus(int indexerId)
+        {
+            return _indexerStatusRepository.FindByIndexerId(indexerId) ?? new IndexerStatus { IndexerId = indexerId };
```
you have a null guard here, but you still check for null all over the place.

Taloth replied:
I think I mentioned this on hangouts. These null checks are needed to make the tests easier.

kayone replied:
core code to make tests easier. _judging_

why not just mock it correctly?

Taloth replied:
That would mean adding it to dozens of testfixtures that shouldn't concern themselves with the indexerstatusservice.
So if you really think it's ugly I'll move the null-coalescing to the places in the service where it's actually used, making the null guards actually required in production code as well.

kayone replied:
I still think it should be responsibility of the test to simulate the context as it would be in production.

kayone commented on src/NzbDrone.Core/Indexers/IndexerFactory.cs:
```diff
@@ -50,12 +50,35 @@ public override IndexerDefinition GetProviderCharacteristics(IIndexer provider,
+            foreach (var indexer in indexers)
+            {
+                var indexerStatus = _indexerStatusService.GetIndexerStatus(indexer.Definition.Id);
+                if (indexerStatus != null && indexerStatus.IsDisabled())
```
don't need the null guard.

kayone commented on src/NzbDrone.Core/Indexers/RssSyncService.cs:
```diff
@@ -62,15 +69,11 @@ private ProcessedDecisions Sync()
 
         public void Execute(RssSyncCommand message)
         {
+            var rssStarted = DateTime.UtcNow;
```
not used?

kayone commented on src/NzbDrone.Core/ThingiProvider/Events/ProviderUpdatedEvent.cs:
```diff
@@ -4,5 +4,11 @@ namespace NzbDrone.Core.ThingiProvider.Events
     {
+        public int ProviderId { get; private set; }
+
+        public ProviderUpdatedEvent(int id)
```
I think this should be the whole definition object rather than just the ID.
